README.md
# `libprimesieve` for Dart

This is a wrapper for [Kim Walisch's `primesieve` C/C++
library](https://github.com/kimwalisch/primesieve).
It requires the library to already be installed.
That library is only available for some platforms, please check
availability.

If you have the underlying native library installed, you can
add this library as `libprimesieve` to your Dart package.

## Looking for a simpler/pure Dart alternative?

If you only need primes up to about 1E9 or prefer to have a Dart-only
implementation, you may want to try
[Prime-Sieve-Dart](https://github.com/MarcelWaldvogel/Prime-Sieve-Dart)
instead, which can be installed with `pub` as `prime-sieve` (note the
dash).

Please note that the pure Dart alternative is significantly slower and
lacks many features.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

# 0.1.2 - 2020-02-12
## Added

## Fixed

## Changed
* Renamed example to make analyzer happy
* Longer description


# 0.1.1 - 2020-02-12
## Added

## Fixed

## Changed
* Package name now `libprimesieve` to avoid conflicts

# 0.1.0 - 2020-02-12
## Added

## Fixed

## Changed
* Created initial version.


test/iterator_test.dart
import 'package:libprimesieve/prime_sequence.dart';
import 'package:libprimesieve/libprimesieve.dart';
import 'package:test/test.dart';

void main() {
  group('Sequences within [1, 1000]', () {
    PrimeSequence ps;

    setUp(() {
      numThreads = 1;
      ps = PrimeSequence(start: 1, stop_hint: 1000);
    });
    tearDown(() {
      ps.freeNativeMemory();
      ps = null;
    });

    test('First prime', () {
      expect(ps.moveNext(), equals(true));
      expect(ps.current, equals(2));
    });
    test('Two primes', () {
      expect(ps.moveNext(), equals(true));
      expect(ps.current, equals(2));
      expect(ps.moveNext(), equals(true));
      expect(ps.current, equals(3));
    });
  });
}


test/primesieve_test.dart
import 'package:libprimesieve/libprimesieve.dart';
import 'package:test/test.dart';
import 'package:path/path.dart' as path;
import 'dart:io';

final execPath = path.join('bin', 'print_primes.dart');

void iotest(List<String> args, String expectStdout) async {
  test('Printing primes with ' + ([execPath] + args).toString(), () async {
    var dartProcess = await Process.run('dart', [execPath] + args);
    expect(dartProcess.exitCode, equals(0));

    expect(dartProcess.stderr, isEmpty);
    expect(dartProcess.stdout, equals(expectStdout));
  });
}

void main() async {
  group('Primes', () {
    test('Primes in [1, 10]', () {
      expect(generatePrimes(1, 10), equals([2, 3, 5, 7]));
    });
    test('10 Primes from 100', () {
      expect(generateNPrimes(10, 100),
          equals([101, 103, 107, 109, 113, 127, 131, 137, 139, 149]));
    });
    test('First prime after 1E9', () {
      expect(nthPrime(1, 1000000000), equals(1000000007));
    });
    test('The prime after 1E9', () {
      expect(nthPrime(0, 1000000000), equals(1000000007));
    });
    test('First prime before 1E9', () {
      expect(nthPrime(-1, 1000000000), equals(999999937));
    });
    test('First prime after 1E9+7 (a prime)', () {
      expect(nthPrime(1, 1000000007), equals(1000000009));
    });
    test('The prime at 1E9+7', () {
      expect(nthPrime(0, 1000000007), equals(1000000007));
    });
    test('First prime before 1E9+7', () {
      expect(nthPrime(-1, 1000000007), equals(999999937));
    });
  });

  group('Big primes (64 bit tests)', () {
    // Not going beyond 2^53 for Javascript
    // Not using 1<<33 etc. for Javascript
    // See https://api.dart.dev/stable/2.7.1/dart-core/int-class.html
    test('First prime before 1E10', () {
      expect(nthPrime(-1, 10000000000), equals(9999999967));
    });
    test('First prime before 1E11', () {
      expect(nthPrime(-1, 100000000000), equals(99999999977));
    });
    test('First prime after 1E12', () {
      expect(nthPrime(1, 100000000000), equals(100000000003));
    });
    test('First prime after 1E15', () {
      expect(nthPrime(1, 100000000000000), equals(100000000000031));
    });
    test('Primes in [1E15, 1E15+100]', () {
      expect(
          generatePrimes(100000000000000, 100000000000100),
          equals([
            100000000000031,
            100000000000067,
            100000000000097,
            100000000000099
          ]));
    });
    test('3 primes starting at 1E15', () {
      expect(generateNPrimes(3, 100000000000000),
          equals([100000000000031, 100000000000067, 100000000000097]));
    });
  });

  group('Counting primes', () {
    test('Count primes in [1, 10]', () {
      expect(countPrimes(1, 10), equals(4));
    });
    test('Count prime twins in [10, 100]', () {
      expect(countTwins(10, 100), equals(6));
    });
    test('Count triplets in [100, 1000]', () {
      expect(countTriplets(100, 1000), equals(21));
    });
    test('Count quadruplets in [1000, 10000]', () {
      expect(countQuadruplets(1000, 10000), equals(7));
    });
    test('Count quintuplets in [10000, 100000]', () {
      expect(countQuintuplets(10000, 100000), equals(12));
    });
    test('Count sextuplets in [100000, 1000000]', () {
      expect(countSextuplets(100000, 1000000), equals(0));
    });
    test('Count sextuplets in [1000000, 10000000]', () {
      expect(countSextuplets(1000000, 10000000), equals(13));
    });
    test('Count sextuplets in [1, 1000000]', () {
      expect(countSextuplets(1, 1000000), equals(5));
    });
    test('Count sextuplets in [1, 10000000]', () {
      expect(countSextuplets(1, 10000000), equals(5 + 13));
    });
  });

  group('Printing primes using example/print_primes.dart', () {
    iotest(['10'], '2\n3\n5\n7\n');
    iotest(['-2', '10'], '(3, 5)\n(5, 7)\n');
    iotest(['-3', '44', '77'], '(67, 71, 73)\n');
    iotest(['-4', '44', '99'], '');
    iotest(['-4', '144', '199'], '(191, 193, 197, 199)\n');
    iotest(['-5', '1000000', '1010000'],
        '(1008851, 1008853, 1008857, 1008859, 1008863)\n');
    iotest(['-6', '2', '100'], '(7, 11, 13, 17, 19, 23)\n');
    iotest(['-16', '7', '23'],
        '7\n11\n13\n\17\n19\n23\n(7, 11, 13, 17, 19, 23)\n');
  });

  group('System configuration', () {
    test('Number of threads', () {
      int n = numThreads;
      expect(n, greaterThan(0));
      expect(n, lessThan(1000));
    });
    test('Modifying numThreads', () {
      int n = numThreads;
      if (n > 1) {
        numThreads = n - 1;
        expect(numThreads, equals(n - 1));
      }
      numThreads = n;
      expect(numThreads, equals(n));
    });
    test('Bounding numThreads', () {
      inhibitWarnings = true; // Do not clobber output with these expected warnings
      int n = numThreads;
      numThreads = 0;
      expect(numThreads, equals(1));
      numThreads = 1000;
      expect(numThreads, equals(n));
      inhibitWarnings = false;
    });
    test('Sieve (wheel) size', () {
      int s = sieveSize;
      expect(s, greaterThanOrEqualTo(7));
      expect(s, lessThan(4096));
    });
    test('Modifying sieveSize', () {
      int s = sieveSize;
      sieveSize = 8;
      expect(sieveSize, equals(8));
      sieveSize = 4096;
      expect(sieveSize, equals(4096));
      sieveSize = s;
      expect(sieveSize, equals(s));
    });
    test('Odd sieveSize', () {
      inhibitWarnings = true;
      int s = sieveSize;
      sieveSize = 15;
      expect(sieveSize, equals(8));
      sieveSize = 3999;
      expect(sieveSize, equals(2048));
      sieveSize = s;
      expect(sieveSize, equals(s));
      inhibitWarnings = false;
    }, tags: ['undocumented', 'unreliable']);
    test('Bounding sieveSize', () {
      inhibitWarnings = true;
      int s = sieveSize;
      sieveSize = 7;
      expect(sieveSize, equals(8));
      sieveSize = 4097;
      expect(sieveSize, equals(4096));
      sieveSize = s;
      expect(sieveSize, equals(s));
      inhibitWarnings = false;
    });
  });

  group('Whether thread count and wheel/cache size affect performance', () {
    test('Reducing number of threads', () {
      int n = numThreads;
      if (n > 1) {
        // Warm up cache
        countPrimes(1, 1000000000);

        // With all cores
        Stopwatch full = Stopwatch()..start();
        countPrimes(1, 1000000000);
        full.stop();

        // With a single core
        numThreads = 1;
        Stopwatch weak = Stopwatch()..start();
        countPrimes(1, 1000000000);
        weak.stop();

        numThreads = n;
        expect(full.elapsed, lessThan(weak.elapsed),
            reason: "$n cores should be faster than 1 core.");
      }
    });

    test('Increasing cache requirements', () {
      int s = sieveSize;
      // Warm up cache
      countPrimes(1, 1000000000);

      // With default cache
      Stopwatch full = Stopwatch()..start();
      countPrimes(1, 1000000000);
      full.stop();

      // With too much cache
      sieveSize = 4096;
      Stopwatch weak = Stopwatch()..start();
      countPrimes(1, 1000000000);
      weak.stop();

      sieveSize = s; // Reset before exception
      expect(full.elapsed, lessThan(weak.elapsed),
          reason: "$s kiB cache should be faster than 4096 kiB.");
    });
  }, tags: ['performance', 'unreliable']);
}


lib/primesieve.dart
// Note that for public APIs directly mapping to library functions,
// minimal Dart functions are created to wrap the `ffi` function.
// This might slightly slow down operation; however, without this workaround,
// API docs created by `dartdoc` would not be helpful.

import 'dart:ffi';
import 'dart:io' show Platform;
import 'package:ffi/ffi.dart';

enum _types {
  SHORT_PRIMES,
  USHORT_PRIMES,
  INT_PRIMES,
  UINT_PRIMES,
  LONG_PRIMES,
  ULONG_PRIMES,
  LONGLONG_PRIMES,
  ULONGLONG_PRIMES,
  INT16_PRIMES,
  UINT16_PRIMES,
  INT32_PRIMES,
  UINT32_PRIMES,
  INT64_PRIMES,
  UINT64_PRIMES
}

// void *primesieve_generate_primes(uint64_t start, uint64_t stop, size_t *size, int type)
typedef _generatePrimesFunc = Pointer<Uint64> Function(
    Uint64 start, Uint64 stop, Pointer<Uint64> size, Uint32 type);
typedef _generatePrimesDart = Pointer<Uint64> Function(
    int start, int stop, Pointer<Uint64> size, int type);

// void *primesieve_generate_n_primes(uint64_t n, uint64_t start, int type)
typedef _generateNPrimesFunc = Pointer<Uint64> Function(
    Uint64 n, Uint64 start, Uint32 type);
typedef _generateNPrimesDart = Pointer<Uint64> Function(
    int n, int start, int type);

// uint64_t primesieve_nth_prime(int64_t n, uint64_t start)
typedef _nthPrimeFunc = Uint64 Function(Uint64 n, Uint64 start);
typedef _nthPrimeDart = int Function(int n, int start);

// uint64_t primesieve_count_primes(uint64_t start, uint64_t stop)
// and friends
typedef _countPrimesFunc = Uint64 Function(Uint64 start, Uint64 stop);
typedef _countPrimesDart = int Function(int start, int stop);

// void primesieve_print_primes(uint64_t start, uint64_t stop)
// and friends
typedef _printPrimesFunc = Void Function(Uint64 start, Uint64 stop);
typedef _printPrimesDart = void Function(int start, int stop);

// void primesieve_free(void *primes)
typedef _freeFunc = Void Function(Pointer<Uint64> primes);
typedef _freeDart = void Function(Pointer<Uint64> primes);

// Generic int getter/setter functions
typedef _getInt64Func = Int64 Function();
typedef _getIntFunc = Int32 Function();
typedef _getIntDart = int Function();
typedef _setIntFunc = Void Function(Int32);
typedef _setIntDart = void Function(int);

typedef _stringFunc = Pointer<Utf8> Function();

final _dylib = _findDL('primesieve');
final _uint64_primes = _types.UINT64_PRIMES.index;

DynamicLibrary _findDL(String basename) {
  final names = _shlibNames(basename);
  for (var name in names) {
    try {
      return DynamicLibrary.open(name);
    } on ArgumentError {}
  }
  throw ArgumentError(
      "Native $basename library not found; searched for $names");
}

List<String> _shlibNames(String basename) {
  if (Platform.isMacOS) return ['lib$basename.dylib'];
  if (Platform.isWindows) return ['$basename.dll'];
  return ['lib$basename.so.9', 'lib$basename.so'];
}

final _free = _dylib
    .lookup<NativeFunction<_freeFunc>>('primesieve_free')
    .asFunction<_freeDart>();
List<int> _uint64arrToIntArr(Pointer<Uint64> arr, int count) {
  // Copy to Dart array
  var primeList = List<int>(count);
  for (var i = 0; i < count; i++) {
    primeList[i] = arr.elementAt(i).value;
  }
  _free(arr);
  return primeList;
}

/// Return an array with all the primes in [[start], [stop]].
///
/// Unless changed with [numThreads], all CPU cores are used.
List<int> generatePrimes(int start, int stop) {
  Pointer<Uint64> sizePointer = allocate();
  final primes = _generatePrimes(start, stop, sizePointer, _uint64_primes);
  final size = sizePointer.value;
  free(sizePointer);
  return _uint64arrToIntArr(primes, size);
}

final _generatePrimes = _dylib
    .lookup<NativeFunction<_generatePrimesFunc>>('primesieve_generate_primes')
    .asFunction<_generatePrimesDart>();

/// Return an array with the first [n] primes >= [start].
///
/// Unless changed with [numThreads], all CPU cores are used.
List<int> generateNPrimes(int n, int start) {
  final primes = _generateNPrimes(n, start, _uint64_primes);
  return _uint64arrToIntArr(primes, n);
}

final _generateNPrimes = _dylib
    .lookup<NativeFunction<_generateNPrimesFunc>>(
        'primesieve_generate_n_primes')
    .asFunction<_generateNPrimesDart>();

/// Find the [n]th prime beyond [start].
///
/// Unless changed with [numThreads], all CPU cores are used.
///
/// Each call to [nthPrime()] incurs an initialization overhead of
/// O(sqrt([start])) even if [n] is tiny. Hence, it is not a good idea to use
/// [nthPrime()] repeatedly in a loop to get the next (or previous) prime.
///
/// For positive/negative [n], the [n]th prime after/before [start] will be
/// returned. For [n] = 0, [start] will be returned, if it is prime, otherwise,
/// the next prime will be returned (i.e., the same prime which [n] = 1 would).
int nthPrime(int n, int start) => _nthPrime(n, start);
final _nthPrime = _dylib
    .lookup<NativeFunction<_nthPrimeFunc>>('primesieve_nth_prime')
    .asFunction<_nthPrimeDart>();

/// Count the primes within the interval [[start], [stop]].
///
/// Unless changed with [numThreads], all CPU cores are used.
///
/// Note that each call to [countPrimes()] incurs an initialization overhead
/// of O(sqrt([stop])) even if the interval [[start], [stop]] is tiny.
int countPrimes(int start, int stop) => _countPrimes(start, stop);
final _countPrimes = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_primes')
    .asFunction<_countPrimesDart>();

/// Count the prime twins within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countTwins(int start, int stop) => _countTwins(start, stop);
final _countTwins = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_twins')
    .asFunction<_countPrimesDart>();

/// Count the prime triplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countTriplets(int start, int stop) => _countTriplets(start, stop);
final _countTriplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_triplets')
    .asFunction<_countPrimesDart>();

/// Count the prime quadruplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countQuadruplets(int start, int stop) => _countQuadruplets(start, stop);
final _countQuadruplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_quadruplets')
    .asFunction<_countPrimesDart>();

/// Count the prime quintuplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countQuintuplets(int start, int stop) => _countQuintuplets(start, stop);
final _countQuintuplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_quintuplets')
    .asFunction<_countPrimesDart>();

/// Count the prime sextuplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countSextuplets(int start, int stop) => _countSextuplets(start, stop);
final _countSextuplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_sextuplets')
    .asFunction<_countPrimesDart>();

/// Print the primes within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printPrimes(int start, int stop) {
  _printPrimes(start, stop);
}

final _printPrimes = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_primes')
    .asFunction<_printPrimesDart>();

/// Print the prime twins within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printTwins(int start, int stop) {
  _printTwins(start, stop);
}

final _printTwins = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_twins')
    .asFunction<_printPrimesDart>();

/// Print the prime triplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printTriplets(int start, int stop) {
  _printTriplets(start, stop);
}

final _printTriplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_triplets')
    .asFunction<_printPrimesDart>();

/// Print the prime quadruplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printQuadruplets(int start, int stop) {
  _printQuadruplets(start, stop);
}

final _printQuadruplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_quadruplets')
    .asFunction<_printPrimesDart>();

/// Print the prime quintuplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printQuintuplets(int start, int stop) {
  _printQuintuplets(start, stop);
}

final _printQuintuplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_quintuplets')
    .asFunction<_printPrimesDart>();

/// Print the prime sextuplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printSextuplets(int start, int stop) {
  _printSextuplets(start, stop);
}

final _printSextuplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_sextuplets')
    .asFunction<_printPrimesDart>();

/// The largest valid `stop` number for primesieve.
///
/// Returns 2^64-1 (`UINT64_MAX`).
int get maxStop => _maxStop();
final _maxStop = _dylib
    .lookup<NativeFunction<_getInt64Func>>('primesieve_get_max_stop')
    .asFunction<_getIntDart>();

/// The sieve size in KiB ([kibibyte](https://en.wikipedia.org/wiki/Kibibyte)).
///
/// Achieve the best sieving performance with a sieve size of your CPU's
/// L1 or L2 cache size (per core).
///
/// [sieveSize] must be at least 8 (8 kiB), at most 4096 (4 MiB).
/// Apparently, the library rounds down sizes to the next power of two.
///
/// When trying to set this to a value modified by the library,
/// a warning will be printed, unless disabled with [inhibitWarnings].
///
/// Like [numThreads], this can be used to tune the performance. However,
/// the default should be reasonable for most settings.
int get sieveSize => _getSieveSize();
set sieveSize(int kiBSize) {
  _setSieveSize(kiBSize);
  if (sieveSize != kiBSize && !inhibitWarnings) {
    print("primesieve.dart: Warning: sieveSize $kiBSize adapted to $sieveSize");
  }
}

final _setSieveSize = _dylib
    .lookup<NativeFunction<_setIntFunc>>('primesieve_set_sieve_size')
    .asFunction<_setIntDart>();
final _getSieveSize = _dylib
    .lookup<NativeFunction<_getIntFunc>>('primesieve_get_sieve_size')
    .asFunction<_getIntDart>();

/// The number of threads to use in all the functions that generate or enumerate
/// primes.
///
/// This defaults to the number of cores and is bounded in [1, number of cores].
/// When trying to set outside this bound, a warning will be printed,
/// unless disabled with [inhibitWarnings].
///
/// Like [sieveSize], this can be used to tune the performance. However,
/// the default should be reasonable for most settings.
int get numThreads => _getNumThreads();
set numThreads(int count) {
  _setNumThreads(count);
  if (numThreads != count && !inhibitWarnings) {
    print("primesieve.dart: Warning: numThreads $count bounded by $numThreads");
  }
}

final _setNumThreads = _dylib
    .lookup<NativeFunction<_setIntFunc>>('primesieve_set_num_threads')
    .asFunction<_setIntDart>();
final _getNumThreads = _dylib
    .lookup<NativeFunction<_getIntFunc>>('primesieve_get_num_threads')
    .asFunction<_getIntDart>();

/// The C/C++ primesieve library version number, in the form “minor.major”.
String get version => Utf8.fromUtf8(_version());
final _version = _dylib
    .lookup<NativeFunction<_stringFunc>>('primesieve_version')
    .asFunction<_stringFunc>();

/// The Dart interface version number, in the form "minor.major".
final dartLibraryVersion = "0.1";

/// Whether to inhibit warnings generated when assigning [numThreads] or
/// [sieveSize] with a value which is modified by the underlying library.
bool inhibitWarnings = false;


lib/libprimesieve.dart
// Note that for public APIs directly mapping to library functions,
// minimal Dart functions are created to wrap the `ffi` function.
// This might slightly slow down operation; however, without this workaround,
// API docs created by `dartdoc` would not be helpful.

import 'dart:ffi';
import 'dart:io' show Platform;
import 'package:ffi/ffi.dart';

enum _types {
  SHORT_PRIMES,
  USHORT_PRIMES,
  INT_PRIMES,
  UINT_PRIMES,
  LONG_PRIMES,
  ULONG_PRIMES,
  LONGLONG_PRIMES,
  ULONGLONG_PRIMES,
  INT16_PRIMES,
  UINT16_PRIMES,
  INT32_PRIMES,
  UINT32_PRIMES,
  INT64_PRIMES,
  UINT64_PRIMES
}

// void *primesieve_generate_primes(uint64_t start, uint64_t stop, size_t *size, int type)
typedef _generatePrimesFunc = Pointer<Uint64> Function(
    Uint64 start, Uint64 stop, Pointer<Uint64> size, Uint32 type);
typedef _generatePrimesDart = Pointer<Uint64> Function(
    int start, int stop, Pointer<Uint64> size, int type);

// void *primesieve_generate_n_primes(uint64_t n, uint64_t start, int type)
typedef _generateNPrimesFunc = Pointer<Uint64> Function(
    Uint64 n, Uint64 start, Uint32 type);
typedef _generateNPrimesDart = Pointer<Uint64> Function(
    int n, int start, int type);

// uint64_t primesieve_nth_prime(int64_t n, uint64_t start)
typedef _nthPrimeFunc = Uint64 Function(Uint64 n, Uint64 start);
typedef _nthPrimeDart = int Function(int n, int start);

// uint64_t primesieve_count_primes(uint64_t start, uint64_t stop)
// and friends
typedef _countPrimesFunc = Uint64 Function(Uint64 start, Uint64 stop);
typedef _countPrimesDart = int Function(int start, int stop);

// void primesieve_print_primes(uint64_t start, uint64_t stop)
// and friends
typedef _printPrimesFunc = Void Function(Uint64 start, Uint64 stop);
typedef _printPrimesDart = void Function(int start, int stop);

// void primesieve_free(void *primes)
typedef _freeFunc = Void Function(Pointer<Uint64> primes);
typedef _freeDart = void Function(Pointer<Uint64> primes);

// Generic int getter/setter functions
typedef _getInt64Func = Int64 Function();
typedef _getIntFunc = Int32 Function();
typedef _getIntDart = int Function();
typedef _setIntFunc = Void Function(Int32);
typedef _setIntDart = void Function(int);

typedef _stringFunc = Pointer<Utf8> Function();

final _dylib = _findDL('primesieve');
DynamicLibrary get dylib => _dylib;
final _uint64_primes = _types.UINT64_PRIMES.index;

DynamicLibrary _findDL(String basename) {
  final names = _shlibNames(basename);
  for (var name in names) {
    try {
      return DynamicLibrary.open(name);
    } on ArgumentError {
      // Try next name now
    }
  }
  throw ArgumentError(
      "Native $basename library not found; searched for $names");
}

List<String> _shlibNames(String basename) {
  if (Platform.isMacOS) return ['lib$basename.dylib'];
  if (Platform.isWindows) return ['$basename.dll'];
  return ['lib$basename.so.9', 'lib$basename.so'];
}

final _free = _dylib
    .lookup<NativeFunction<_freeFunc>>('primesieve_free')
    .asFunction<_freeDart>();
List<int> _convertAndFreeNativePrimeArray(Pointer<Uint64> arr, int count) {
  // Copy to Dart array
  var primeList = List<int>(count);
  for (var i = 0; i < count; i++) {
    primeList[i] = arr.elementAt(i).value;
  }
  _free(arr);
  return primeList;
}

/// Return an array with all the primes in [[start], [stop]].
///
/// Unless changed with [numThreads], all CPU cores are used.
List<int> generatePrimes(int start, int stop) {
  Pointer<Uint64> sizePointer = allocate();
  final primes = _generatePrimes(start, stop, sizePointer, _uint64_primes);
  final size = sizePointer.value;
  free(sizePointer);
  return _convertAndFreeNativePrimeArray(primes, size);
}

final _generatePrimes = _dylib
    .lookup<NativeFunction<_generatePrimesFunc>>('primesieve_generate_primes')
    .asFunction<_generatePrimesDart>();

/// Return an array with the first [n] primes >= [start].
///
/// Unless changed with [numThreads], all CPU cores are used.
List<int> generateNPrimes(int n, int start) {
  final primes = _generateNPrimes(n, start, _uint64_primes);
  return _convertAndFreeNativePrimeArray(primes, n);
}

final _generateNPrimes = _dylib
    .lookup<NativeFunction<_generateNPrimesFunc>>(
        'primesieve_generate_n_primes')
    .asFunction<_generateNPrimesDart>();

/// Find the [n]th prime beyond [start].
///
/// Unless changed with [numThreads], all CPU cores are used.
///
/// Each call to [nthPrime()] incurs an initialization overhead of
/// O(sqrt([start])) even if [n] is tiny. Hence, it is not a good idea to use
/// [nthPrime()] repeatedly in a loop to get the next (or previous) prime.
///
/// For positive/negative [n], the [n]th prime after/before [start] will be
/// returned. For [n] = 0, [start] will be returned, if it is prime, otherwise,
/// the next prime will be returned (i.e., the same prime which [n] = 1 would).
int nthPrime(int n, int start) => _nthPrime(n, start);
final _nthPrime = _dylib
    .lookup<NativeFunction<_nthPrimeFunc>>('primesieve_nth_prime')
    .asFunction<_nthPrimeDart>();

/// Count the primes within the interval [[start], [stop]].
///
/// Unless changed with [numThreads], all CPU cores are used.
///
/// Note that each call to [countPrimes()] incurs an initialization overhead
/// of O(sqrt([stop])) even if the interval [[start], [stop]] is tiny.
int countPrimes(int start, int stop) => _countPrimes(start, stop);
final _countPrimes = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_primes')
    .asFunction<_countPrimesDart>();

/// Count the prime twins within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countTwins(int start, int stop) => _countTwins(start, stop);
final _countTwins = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_twins')
    .asFunction<_countPrimesDart>();

/// Count the prime triplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countTriplets(int start, int stop) => _countTriplets(start, stop);
final _countTriplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_triplets')
    .asFunction<_countPrimesDart>();

/// Count the prime quadruplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countQuadruplets(int start, int stop) => _countQuadruplets(start, stop);
final _countQuadruplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_quadruplets')
    .asFunction<_countPrimesDart>();

/// Count the prime quintuplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countQuintuplets(int start, int stop) => _countQuintuplets(start, stop);
final _countQuintuplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_quintuplets')
    .asFunction<_countPrimesDart>();

/// Count the prime sextuplets within the interval [[start], [stop]].
///
/// See [countPrimes()] for more information.
int countSextuplets(int start, int stop) => _countSextuplets(start, stop);
final _countSextuplets = _dylib
    .lookup<NativeFunction<_countPrimesFunc>>('primesieve_count_sextuplets')
    .asFunction<_countPrimesDart>();

/// Print the primes within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printPrimes(int start, int stop) {
  _printPrimes(start, stop);
}

final _printPrimes = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_primes')
    .asFunction<_printPrimesDart>();

/// Print the prime twins within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printTwins(int start, int stop) {
  _printTwins(start, stop);
}

final _printTwins = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_twins')
    .asFunction<_printPrimesDart>();

/// Print the prime triplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printTriplets(int start, int stop) {
  _printTriplets(start, stop);
}

final _printTriplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_triplets')
    .asFunction<_printPrimesDart>();

/// Print the prime quadruplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printQuadruplets(int start, int stop) {
  _printQuadruplets(start, stop);
}

final _printQuadruplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_quadruplets')
    .asFunction<_printPrimesDart>();

/// Print the prime quintuplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printQuintuplets(int start, int stop) {
  _printQuintuplets(start, stop);
}

final _printQuintuplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_quintuplets')
    .asFunction<_printPrimesDart>();

/// Print the prime sextuplets within the interval [[start], [stop]].
///
/// See [printPrimes()] for more information.
void printSextuplets(int start, int stop) {
  _printSextuplets(start, stop);
}

final _printSextuplets = _dylib
    .lookup<NativeFunction<_printPrimesFunc>>('primesieve_print_sextuplets')
    .asFunction<_printPrimesDart>();

/// The largest valid `stop` number for primesieve.
///
/// Returns 2^64-1 (`UINT64_MAX`).
int get maxStop => _maxStop();
final _maxStop = _dylib
    .lookup<NativeFunction<_getInt64Func>>('primesieve_get_max_stop')
    .asFunction<_getIntDart>();

/// The sieve size in KiB ([kibibyte](https://en.wikipedia.org/wiki/Kibibyte)).
///
/// Achieve the best sieving performance with a sieve size of your CPU's
/// L1 or L2 cache size (per core).
///
/// [sieveSize] must be at least 8 (8 kiB), at most 4096 (4 MiB).
/// Apparently, the library rounds down sizes to the next power of two.
///
/// When trying to set this to a value modified by the library,
/// a warning will be printed, unless disabled with [inhibitWarnings].
///
/// Like [numThreads], this can be used to tune the performance. However,
/// the default should be reasonable for most settings.
int get sieveSize => _getSieveSize();
set sieveSize(int kiBSize) {
  _setSieveSize(kiBSize);
  if (sieveSize != kiBSize && !inhibitWarnings) {
    print("primesieve.dart: Warning: sieveSize $kiBSize adapted to $sieveSize");
  }
}

final _setSieveSize = _dylib
    .lookup<NativeFunction<_setIntFunc>>('primesieve_set_sieve_size')
    .asFunction<_setIntDart>();
final _getSieveSize = _dylib
    .lookup<NativeFunction<_getIntFunc>>('primesieve_get_sieve_size')
    .asFunction<_getIntDart>();

/// The number of threads to use in all the functions that generate or enumerate
/// primes.
///
/// This defaults to the number of cores and is bounded in [1, number of cores].
/// When trying to set outside this bound, a warning will be printed,
/// unless disabled with [inhibitWarnings].
///
/// Like [sieveSize], this can be used to tune the performance. However,
/// the default should be reasonable for most settings.
int get numThreads => _getNumThreads();
set numThreads(int count) {
  _setNumThreads(count);
  if (numThreads != count && !inhibitWarnings) {
    print("primesieve.dart: Warning: numThreads $count bounded by $numThreads");
  }
}

final _setNumThreads = _dylib
    .lookup<NativeFunction<_setIntFunc>>('primesieve_set_num_threads')
    .asFunction<_setIntDart>();
final _getNumThreads = _dylib
    .lookup<NativeFunction<_getIntFunc>>('primesieve_get_num_threads')
    .asFunction<_getIntDart>();

/// The C/C++ primesieve library version number, in the form “minor.major”.
String get version => Utf8.fromUtf8(_version());
final _version = _dylib
    .lookup<NativeFunction<_stringFunc>>('primesieve_version')
    .asFunction<_stringFunc>();

/// The Dart interface version number, in the form "minor.major".
final dartLibraryVersion = "0.1";

/// Whether to inhibit warnings generated when assigning [numThreads] or
/// [sieveSize] with a value which is modified by the underlying library.
bool inhibitWarnings = false;
